qiita_db/software.py

```diff
@@ -0,0 +1,522 @@
+            if cls.exists(command, **kwargs):
+                raise QiitaDBDuplicateError(command.parameters_table,
+                                            "Values: %s" % kwargs)
+
+            vals = kwargs.values()
+            vals.insert(0, param_set_name)
+
+            sql = """INSERT INTO qiita.{0} (param_set_name, {1})
+                     VALUES (%s, {2})
+                     RETURNING parameters_id""".format(
+                command.parameters_table,
+                ', '.join(kwargs),
+                ', '.join(['%s'] * len(kwargs)))
+            TRN.add(sql, vals)
+
+            return cls(TRN.execute_fetchlast(), command)
+
+    @classmethod
+    def iter(cls, command):
+        """Iterates over all parameter sets of the given command
+
+        Returns
+        -------
+        generator
+            Yields a parameter instance
+        """
+        with TRN:
+            sql = """SELECT parameters_id
+                     FROM qiita.{0}
+                     ORDER BY parameters_id""".format(
+                command.parameters_table)
+            TRN.add(sql)
+            for result in TRN.execute_fetchflatten():
+                yield cls(result, command)
+
+    @property
+    def name(self):
+        """The name of the parameter set
+
+        Returns
+        -------
+        str
+            The name of the parameter set
+        """
+        with TRN:
+            sql = """SELECT param_set_name
+                     FROM qiita.{0}
+                     WHERE parameters_id = %s""".format(self._table)
+            TRN.add(sql, [self.id])
+            return TRN.execute_fetchlast()
+
+    @property
+    def values(self):
+        """The values of the parameter set
+
+        Returns
+        -------
+        dict
+            Dictionary with the parameters values keyed by parameter name
+        """
+        with TRN:
+            sql = "SELECT * FROM qiita.{0} WHERE parameters_id = %s".format(
+                self._table)
+            TRN.add(sql, [self.id])
+            # There should be only one row
+            result = dict(TRN.execute_fetchindex()[0])
+            del result["parameters_id"]
+            del result["param_set_name"]
+            return result
+
+    def to_str(self):
+        """Generates a string with the parameter values
+
+        Returns
+        -------
+        str
+            The string with all the parameters
+        """
+        with TRN:
+            table_cols = get_table_cols_w_type(self._table)
+            table_cols.remove(['parameters_id', 'bigint'])
+            table_cols.remove(['param_set_name', 'character varying'])
+
+            values = self.values
+
+            result = []
+            for p_name, p_type in sorted(table_cols):
+                if p_type == 'boolean':
+                    if values[p_name]:
+                        result.append("--%s" % p_name)
+                else:
+                    result.append("--%s %s" % (p_name, values[p_name]))
```

Comment:
Would it be useful to always quote the arguments? Not sure if that would bite us somehow, but when we've had command-calls before, its been useful to quote arguments in case one of them includes spaces. Which although chaotic and unforgivable, tend to inevitably happen.